### PR TITLE
Binary targets workspace implementation

### PR DESCRIPTION
**Pull request by hartbit**
This implements the download and unarchiving implementation of the binary targets proposal.

### CODE REVIEW

File: Sources/Commands/SwiftTool.swift [hunk 1]
```diff
@@ -142,6 +159,31 @@ private class ToolWorkspaceDelegate: WorkspaceDelegate {
...
+            }
+
+            let step = self.downloadProgress.values.reduce(0, { $0 + $1.bytesDownloaded }) / 1024
+            let total = self.downloadProgress.values.reduce(0, { $0 + $1.totalBytesToDownload }) / 1024
+            self.downloadAnimation.update(step: Int(step), total: Int(total), text: "Downloading binary artifacts...")
+        }
+    }
+
+    func didDownloadBinaryArtifacts() {
+        queue.async {
```
Comment (aciidb0mb3r) [inline Sources/Commands/SwiftTool.swift]: Do we need to add `queue.sync {}` in the deinit for this class so we don't end up exiting while we have things in the queue?

File: Sources/Workspace/Workspace.swift [hunk 2]
```diff
@@ -420,6 +441,7 @@ public class Workspace {
         self.repositoryManager = repositoryManager
 
         self.checkoutsPath = self.dataPath.appending(component: "checkouts")
+        self.artifactsPath = self.dataPath.appending(component: "artifacts")
```
Comment (aciidb0mb3r) [inline Sources/Workspace/Workspace.swift]: can you move this to be inside the checkouts directory.

Comment (hartbit) [reply Sources/Workspace/Workspace.swift]: That will clash with a SwiftPackage named `artifacts`, won't it?

File: Sources/Commands/SwiftTool.swift (hunk 1 above)
Comment (hartbit) [reply Sources/Commands/SwiftTool.swift]: I don't think its necessary. The async closure implicitly captures self, so `ToolWorkspaceDelegate` can't be deinit before all those closures finish running.

**Comment by hartbit** [comment]
I have several diagnostics in the PR reported using the `emit(error: "description of error")` instead of `emit(.errorName(arguments))`. I thought the second form was preferred, but I also saw a few instances of the first form using in Workspace. @aciidb0mb3r what do you think?

File: Sources/Workspace/Workspace.swift (hunk 2 above)
Comment (aciidb0mb3r) [reply Sources/Workspace/Workspace.swift]: Ah because we dropped the suffix in the names. I guess this path is also fine.

**Comment by aciidb0mb3r** [comment]
@hartbit The string variant is ok for one-off cases where we don't need to emit the same error from somewhere else but `emit(.errorName(arguments))` is probably better/preferable.

File: Sources/PackageLoading/PackageBuilder.swift [hunk 3]
```diff
@@ -619,6 +651,15 @@ public final class PackageBuilder {
                 pkgConfig: manifestTarget.pkgConfig,
                 providers: manifestTarget.providers
             )
+        } else if potentialModule.type == .binary {
+            let remoteURL = remoteArtifacts.first(where: { $0.path.basenameWithoutExt == potentialModule.name })
```
Comment (aciidb0mb3r) [inline Sources/PackageLoading/PackageBuilder.swift]: It might be better to perform this lookup using the artifact path.

**Comment by hartbit** [comment]
@swift-ci smoke test self hosted

**Comment by hartbit** [comment]
@swift-ci smoke test

**Pull request merged**